PM: Add constraint API

ceolin opened the pull request:
This will replace the pm_ctrl_* API.

pabigot commented on subsys/power/power.c:
```diff
@@ -198,6 +204,58 @@ void pm_system_resume(void)
+
+	k_mutex_lock(&pm_constraint_lock, K_FOREVER);
+
+	if (size_add_overflow(pm_constraints[state], 1,
```
I'm really not sure this is worth doing.  Yes, it's technically possible for somebody to set the same constraint 2^31 times, and this would prevent that from introducing undefined behavior.

However, it also means that we should have an expectation that people will not on check results for errors but be able to react in some reasonable way if they occur.

pabigot commented on include/power/power.h:
```diff
@@ -314,6 +315,44 @@ void pm_notify_power_state_entry(enum power_states state);
+/**
+ * @brief Set constraint for a power states.
+ *
+ * A constraint will prevents the system to go to a specific power state.
```
Document that setting and releasing constraints must be done from a thread context.

pabigot reviewed (commented):
Before we go too far down the path of size-checked increments I'd like to see that discussed.

Also if we're using a mutex we need to document that people can't set constraints from interrupts or before `POST_KERNEL`.  Which, I suspect, is probably a restriction we should not impose.

ceolin commented:
@pabigot thanks for reviewing it. Regarding mutex usage, I can remove them if I assume that the input is always right, but I don't like this idea, right now this is just checking if the given state is a valid one, but we may want to check if the state is defined by the target. I can change it to spin look and allows using from an interrupt context.

Why do you think it is bad constraint allowing use this API only after `POST_KERNEL` ?

pabigot commented:
> Why do you think it is bad constraint allowing use this API only after `POST_KERNEL` ?

When drivers start up pre-kernel they may want to set constraints to prevent themselves from being shut down.

Commit:
power: Rename constraint API

Replace pm_ctrl_* with pm_constraint.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Remove PM_STATE_LOCK option

Simplify pm subsystem removing PM_STATE_LOCK option. Constraints API is
small and is a key component of power subsystem.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: constraints: Update constraint documentation

Align documentation with constraint API changes.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Fix documentation

Small fixes in state control documentation.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@pabigot I just  kept the current implementation but changed API names to align with what you've discussed. Would you mind take a look again ?

pabigot reviewed (approved):
I continue to believe the documentation on this API is horrifically poor (what does "set a constraint" actually mean in terms of its effect on the system behavior?), but as a refactoring PR it's ok.

ceolin commented:
@nashif would you mind take a look ? Also, CI failed for not clear reason.

ceolin commented:
> I continue to believe the documentation on this API is horrifically poor (what does "set a constraint" actually mean in terms of its effect on the system behavior?), but as a refactoring PR it's ok.

agree, will address it in another pr.

andyross reviewed (approved):
Seems fine.  I agree that naming cleanup without a doc rewrite might be getting ahead of ourselves.  But no reason not to merge.

erwango commented on include/power/power.h:
```diff
@@ -159,31 +159,34 @@ void pm_dump_debug_info(void);
-#ifdef CONFIG_PM_STATE_LOCK
 /**
- * @brief Disable particular power state
+ * @brief Set a constraint for a power state
```
One point is not clear to me on this API. Not sure if this is the right pace to discuss. Please discard this comment if this isn't.
When setting a constraint on a particular state, should we assume the same constraint should be used on "deeper sleep states", or is the caller expected to call all related states where the constrain apply?

What about substates ?

pabigot replied:
I think we need to wait to discuss that, since this PR is just renaming stuff, but:

* I'm not sure if we can assume a strict linear ordering such that if you block a transition to `STANDBY` you implicitly block transition to `SUSPEND_TO_RAM` and other "lower" states.  Need to see what the prior art does.
* yes, constraints are going to have to take substates into account, either explicitly or through some other back door.
